# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/xpconnect/src/XPCWrappedNativeJSOps.cpp
# Commit: b3c093b141e2
# Full Hash: b3c093b141e2381efef699124f8a629d0711c8c6
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2018-08-07 15:42:55
# Regressor Bug: 1478359
# File Overlap Count: 2
# Description:
#   Bug 1478359 - Store a global object in nsXPCWrappedJS and use it for realm-entering. r=mccr8
#   
#   The problem we're solving here: getting/entering the realm/global of a cross-compartment wrapper doesn't make sense once there are multiple realms in a compartment and the CCW will be shared by all of them. Because nsXPCWrappedJS can store a CCW, we will no longer be able to use this JSObject to enter the target realm.
#   
#   What this patch does: we pass a JSContext* to nsXPCWrappedJS::GetNewOrUsed and we use this to store a global object in nsXPCWrappedJS (with the invariant that the object and global stored in nsXPCWrappedJS are same-compartment). Then when we want to enter the nsXPCWrappedJS's target realm, we use this global object instead of the maybe-CCW object. Because we currently still have one realm per compartment and the objects are same-compartment, this is guaranteed to preserve behavior for now.
# ==============================================================================

diff -r bf6961db9405 -r b3c093b141e2 js/xpconnect/src/XPCWrappedNativeJSOps.cpp
--- a/js/xpconnect/src/XPCWrappedNativeJSOps.cpp	Tue Aug 07 12:41:15 2018 +0300
+++ b/js/xpconnect/src/XPCWrappedNativeJSOps.cpp	Tue Aug 07 11:57:41 2018 +0200
@@ -157,7 +157,7 @@
             RootedId id(ccx, ccx.GetContext()->
                             GetStringID(XPCJSContext::IDX_WRAPPED_JSOBJECT));
 
-            JSAutoRealmAllowCCW ar(ccx, mainObj);
+            JSAutoRealm ar(ccx, underware->GetJSObjectGlobal());
 
             RootedValue val(ccx);
             if (JS_GetPropertyById(ccx, mainObj, id, &val) &&